crates/cairo-profiler/src/profile_viewer.rs

```diff
@@ -0,0 +1,168 @@
+use anyhow::{bail, Context, Result};
+use camino::Utf8PathBuf;
+use flate2::read::GzDecoder;
+use prettytable::{format, Table};
+use prost::Message;
+use std::collections::HashMap;
+use std::fs;
+use std::io::Read;
+
+use crate::profile_builder::pprof::{Function, Location, Profile};
+
+#[derive(Debug, Default)]
+struct FunctionProfile {
+    flat: i64,
+    flat_p: f64,
+    cumulative: i64,
+    cumulative_p: f64,
+    sum_p: f64,
+}
+
+fn get_profile_data(
+    profile: &Profile,
+    sample_name: &str,
+) -> Result<Vec<(String, FunctionProfile)>> {
+    // Labels in string_table are prefixed with a whitespace
+    let sample_label = format!(" {sample_name}");
+
+    let sample_type_idx = profile
+        .sample_type
+        .iter()
+        .position(|sample| {
+            profile.string_table[usize::try_from(sample.unit)
+                .expect("Overflow while converting samples id to usize")]
+                == sample_label
+        })
+        .context("Failed to find sample in provided profile")?;
+
+    let mut profile_map = HashMap::<String, FunctionProfile>::new();
+
+    let location_map: HashMap<u64, &Location> = profile
+        .location
+        .iter()
+        .map(|location| (location.id, location))
+        .collect();
+
+    let function_map: HashMap<u64, &Function> = profile
+        .function
+        .iter()
+        .map(|function| (function.id, function))
+        .collect();
+
+    for sample in &profile.sample {
+        let sample_value = sample.value[sample_type_idx];
+
+        sample
+            .location_id
+            .iter()
+            .filter_map(|&loc_id| {
+                location_map
```

Comment:
What about something like this:

```rs
let line = location_map.get(&loc_id)?.line.first()?;
let function = function_map.get(&line.function_id)?;

Some(
    &profile.string_table[usize::try_from(function.name)
     .expect("Overflow while converting function name id to usize")],
)
```

Reply:
Hmm, I guess a (minor) perf hit is worth the readability, and it looks more elegant. Changed :+1: